macOS/unwindmacos.cpp

```diff
@@ -0,0 +1,210 @@
+*/
+
+#include "unwindmacos.hpp"
+
+#include <mach-o/dyld.h>
+#include <mach-o/dyld_images.h>
+#include <mach-o/nlist.h>
+
+UnwindDataMacOS::UnwindDataMacOS() {
+  addr = NULL;
+  size = 0;
+  buffer = NULL;
+  header = NULL;
+}
+
+void UnwindDataMacOS::AddEncoding(compact_unwind_encoding_t encoding, size_t translated_address) {
+  if (metadata_list.empty() || metadata_list.back().encoding != encoding) {
+    metadata_list.push_back(Metadata(encoding, translated_address, translated_address));
+  }
+  else {
+    metadata_list.back().max_address = translated_address;
+  }
+}
+
+UnwindGeneratorMacOS::UnwindGeneratorMacOS(TinyInst& tinyinst) : UnwindGenerator(tinyinst) {
+
+}
+
+
+void UnwindGeneratorMacOS::OnModuleInstrumented(ModuleInfo* module) {
+  size_t file_vm_slide = 0;
+  section_64 unwind_section;
+  tinyinst_.GetSectionAndSlide(module->module_header, "__TEXT", "__unwind_info", &unwind_section, &file_vm_slide);
+
+  module->unwind_data = new UnwindDataMacOS();
+  module->unwind_data->addr = (void*)(file_vm_slide + unwind_section.addr);
+  module->unwind_data->size = unwind_section.size;
+
+  module->unwind_data->buffer = (void*)malloc(unwind_section.size);
+  tinyinst_.RemoteRead(module->unwind_data->addr,
+                       module->unwind_data->buffer,
+                       module->unwind_data->size);
+
+  module->unwind_data->header = (struct unwind_info_section_header *)module->unwind_data->buffer;
+}
+
+void UnwindGeneratorMacOS::OnBasicBlockStart(ModuleInfo* module,
+                                             size_t original_address,
+                                             size_t translated_address) {
+  FirstLevelLookup(module, original_address, translated_address);
+}
+
+void UnwindGeneratorMacOS::OnInstruction(ModuleInfo* module,
+                                         size_t original_address,
+                                         size_t translated_address) {
+  FirstLevelLookup(module, original_address, translated_address);
+}
+
+void UnwindGeneratorMacOS::OnBasicBlockEnd(ModuleInfo* module,
+                                           size_t original_address,
+                                           size_t translated_address) {
+  FirstLevelLookup(module, original_address, translated_address - 1);
+}
+
+void UnwindGeneratorMacOS::OnModuleUninstrumented(ModuleInfo *module) {
+  free(module->unwind_data->buffer);
+  delete module->unwind_data;
+}
+
+void UnwindGeneratorMacOS::FirstLevelLookup(ModuleInfo *module, size_t original_address, size_t translated_address) {
+  size_t original_offset = original_address - (size_t)module->module_header;
+
+  struct unwind_info_section_header_index_entry *first_level_entry = NULL;
+  size_t first_level_entries_start = (size_t)module->unwind_data->buffer
+                                      + module->unwind_data->header->indexSectionOffset;
+
+  uint32_t low_index = 0;
+  uint32_t high_index = module->unwind_data->header->indexCount;
+  while (low_index < high_index) {
```

Comment:
I'm not sure this will ever happen but think about how to detect / handle the case when indexCount is 0 or a corresponding entry is not found in the search.